driver/gather_driver.hpp

```diff
@@ -0,0 +1,389 @@
+#include "mloGatherHost.hpp"
+#include "random.hpp"
+#include "tensor_driver.hpp"
+#include "timer.hpp"
+
+#include <../test/verify.hpp>
+
+#include <miopen/errors.hpp>
+#include <miopen/miopen.h>
+#include <miopen/tensor_view_utils.hpp>
+
+#include <memory>
+#include <vector>
+
+template <typename Tgpu, typename Tref, typename Tindex>
+class GatherDriver : public Driver
+{
+public:
+    GatherDriver() : Driver()
+    {
+        miopenCreateTensorDescriptor(&inputTensor);
+        miopenCreateTensorDescriptor(&indicesTensor);
+        miopenCreateTensorDescriptor(&outputTensor);
+
+        data_type       = miopen_type<Tgpu>{};
+        index_data_type = miopen_type<Tindex>{};
+    }
+
+    int AddCmdLineArgs() override;
+    int ParseCmdLineArgs(int argc, char* argv[]) override;
+    InputFlags& GetInputFlags() override { return inflags; }
+
+    std::vector<int> ComputeStrides(std::vector<int> inputDim);
+
+    int GetandSetData() override;
+
+    int AllocateBuffersAndCopy() override;
+
+    int RunForwardGPU() override;
+    int RunForwardCPU(); // Verify implements it
+
+    int RunBackwardGPU() override;
+    int RunBackwardCPU(); // Verify implements it
+
+    Tref GetTolerance();
+    int VerifyBackward() override;
+    int VerifyForward() override;
+    ~GatherDriver() override
+    {
+        miopenDestroyTensorDescriptor(inputTensor);
+        miopenDestroyTensorDescriptor(indicesTensor);
+        miopenDestroyTensorDescriptor(outputTensor);
+    }
+
+private:
+    int forw;
+    bool isContiguous;
+
+    InputFlags inflags;
+    miopenDataType_t index_data_type;
+    miopenGatherDescriptor_t gatherDesc;
+
+    miopenTensorDescriptor_t inputTensor;
+    miopenTensorDescriptor_t indicesTensor;
+    miopenTensorDescriptor_t outputTensor;
+
+    std::unique_ptr<GPUMem> input_dev;
+    std::unique_ptr<GPUMem> indices_dev;
+    std::unique_ptr<GPUMem> output_dev;
+
+    std::vector<Tgpu> input;
+    std::vector<Tindex> indices;
+    std::vector<Tgpu> output;
+
+    std::vector<Tref> outputHost;
+
+    miopenGatherMode_t mode;
+    uint32_t dim;
+    uint32_t batch_dims;
+};
+
+// Equivalent tensor.transpose(0, -1).contiguous().transpose(0, -1)
+template <typename Tgpu, typename Tref, typename Tindex>
+std::vector<int> GatherDriver<Tgpu, Tref, Tindex>::ComputeStrides(std::vector<int> inputDim)
+{
+    if(!isContiguous)
+        std::swap(inputDim.front(), inputDim.back());
+    std::vector<int> strides(inputDim.size());
+    strides.back() = 1;
+    for(int i = inputDim.size() - 2; i >= 0; --i)
+        strides[i] = strides[i + 1] * inputDim[i + 1];
+    if(!isContiguous)
+        std::swap(strides.front(), strides.back());
+    return strides;
+}
+
+template <typename Tgpu, typename Tref, typename Tindex>
+int GatherDriver<Tgpu, Tref, Tindex>::ParseCmdLineArgs(int argc, char* argv[])
+{
+    inflags.Parse(argc, argv);
+
+    forw = inflags.GetValueInt("forw");
+    MIOPEN_THROW_IF(forw != 0 && forw != 1, "Incorrect Forward Mode");
```

Comment:
missed this `MIOPEN_THROW_IF(forw != 0 || forw != 1, "Incorrect Forward Mode");`